Volume remain stuck in pending state

**Describe the bug**
Volume remain stuck in pending state and additionally `RESTORESIZE` field is empty for `VolumeSnapshot` Object and panic is observed in cephlet

**To Reproduce**
1. Create the Volume using below manifest:
```apiVersion: storage.api.onmetal.de/v1alpha1
kind: Volume
metadata:
  name: machine-stefan-2
spec:
  volumePoolRef:
    name: ceph
  volumeClassRef:
    name: super
  image: ghcr.io/hardikdr/onmetal-image/gardenlinux:ign-test8
  resources:
    storage: 15Gi
    ```

**Expected behavior**
Volume should be created with Available state and no panic should be observed

**Screenshots**


Cephlet logs:
```2023-02-15T10:54:07Z	INFO	Found datasource ref for PVC	{"controller": "persistentvolumeclaim", "controllerGroup": "", "controllerKind": "PersistentVolumeClaim", "PersistentVolumeClaim": {"name":"ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8", "reconcileID": "9d66c3f3-a806-4f36-a92e-3213ccdd747e", "DataSourceRef": "&TypedObjectReference{APIGroup:*storage.api.onmetal.de/v1alpha1,Kind:Volume,Name:volume-ap,Namespace:nil,}"}
2023-02-15T10:54:07Z	INFO	Requeue: Datasource ref for PVC could not be found	{"controller": "persistentvolumeclaim", "controllerGroup": "", "controllerKind": "PersistentVolumeClaim", "PersistentVolumeClaim": {"name":"ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8", "reconcileID": "9d66c3f3-a806-4f36-a92e-3213ccdd747e", "datasource": {"namespace": "rook-ceph", "name": "volume-ap"}}






2023-02-15T10:54:43Z	DEBUG	Reconciling Volume	{"controller": "volume", "controllerGroup": "storage.api.onmetal.de", "controllerKind": "Volume", "Volume": {"name":"volume-ap","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "volume-ap", "reconcileID": "2320d30c-f559-43a1-8620-00b3a4ee8d74"}
2023-02-15T10:54:43Z	DEBUG	Requested snapshot not found: create image pvc and snapshot it.	{"controller": "volume", "controllerGroup": "storage.api.onmetal.de", "controllerKind": "Volume", "Volume": {"name":"volume-ap","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "volume-ap", "reconcileID": "2320d30c-f559-43a1-8620-00b3a4ee8d74"}
2023-02-15T10:54:45Z	DEBUG	created image pvc rook-ceph/ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8	{"controller": "volume", "controllerGroup": "storage.api.onmetal.de", "controllerKind": "Volume", "Volume": {"name":"volume-ap","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "volume-ap", "reconcileID": "2320d30c-f559-43a1-8620-00b3a4ee8d74"}
2023-02-15T10:54:45Z	INFO	Reconciling PVC	{"controller": "persistentvolumeclaim", "controllerGroup": "", "controllerKind": "PersistentVolumeClaim", "PersistentVolumeClaim": {"name":"ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8", "reconcileID": "87f81523-d21a-481c-a53e-ba8ac7d8d7ac"}
2023-02-15T10:54:45Z	INFO	Found datasource ref for PVC	{"controller": "persistentvolumeclaim", "controllerGroup": "", "controllerKind": "PersistentVolumeClaim", "PersistentVolumeClaim": {"name":"ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8", "reconcileID": "87f81523-d21a-481c-a53e-ba8ac7d8d7ac", "DataSourceRef": "&TypedObjectReference{APIGroup:*storage.api.onmetal.de/v1alpha1,Kind:Volume,Name:volume-ap,Namespace:nil,}"}
2023-02-15T10:54:45Z	INFO	Found volume as datasource ref for PVC	{"controller": "persistentvolumeclaim", "controllerGroup": "", "controllerKind": "PersistentVolumeClaim", "PersistentVolumeClaim": {"name":"ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8", "reconcileID": "87f81523-d21a-481c-a53e-ba8ac7d8d7ac", "Volume": {"namespace": "rook-ceph", "name": "volume-ap"}}
2023-02-15T10:54:45Z	INFO	Found StorageClass for PVC	{"controller": "persistentvolumeclaim", "controllerGroup": "", "controllerKind": "PersistentVolumeClaim", "PersistentVolumeClaim": {"name":"ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8", "reconcileID": "87f81523-d21a-481c-a53e-ba8ac7d8d7ac", "StorageClass": {"name": "volume-rook-ceph--ceph"}}
2023-02-15T10:54:45Z	DEBUG	created snapshot rook-ceph/ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8 for image pvc rook-ceph/ghcr.io-hardikdr-onmetal-image-gardenlinux-ign-test8	{"controller": "volume", "controllerGroup": "storage.api.onmetal.de", "controllerKind": "Volume", "Volume": {"name":"volume-ap","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "volume-ap", "reconcileID": "2320d30c-f559-43a1-8620-00b3a4ee8d74"}
2023-02-15T10:54:45Z	DEBUG	Pvc is not yet in ClaimBound state	{"controller": "volume", "controllerGroup": "storage.api.onmetal.de", "controllerKind": "Volume", "Volume": {"name":"volume-ap","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "volume-ap", "reconcileID": "2320d30c-f559-43a1-8620-00b3a4ee8d74"}
2023-02-15T10:54:45Z	DEBUG	Reconciling Volume	{"controller": "volume", "controllerGroup": "storage.api.onmetal.de", "controllerKind": "Volume", "Volume": {"name":"volume-ap","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "volume-ap", "reconcileID": "6578d357-2e8d-4b36-b9e9-149c6792695c"}
2023-02-15T10:54:45Z	INFO	Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference	{"controller": "volume", "controllerGroup": "storage.api.onmetal.de", "controllerKind": "Volume", "Volume": {"name":"volume-ap","namespace":"rook-ceph"}, "namespace": "rook-ceph", "name": "volume-ap", "reconcileID": "6578d357-2e8d-4b36-b9e9-149c6792695c"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1720d8f]

goroutine 440 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:119 +0x1fa
panic({0x18fea40, 0x2ae0a10})
	/usr/local/go/src/runtime/panic.go:884 +0x213
github.com/onmetal/cephlet/controllers.(*VolumeReconciler).handleImagePopulation(0xc00002a9c0, {0x1de04f8, 0xc000c26570}, {{0x1de2ff0?, 0xc000c265a0?}, 0x0?}, 0xc000c48000)
	/workspace/controllers/volume_controller.go:360 +0x28f
github.com/onmetal/cephlet/controllers.(*VolumeReconciler).reconcile(0xc00002a9c0, {0x1de04f8, 0xc000c26570}, {{0x1de2ff0?, 0xc000c265a0?}, 0x1df4d30?}, 0xc000c48000)
	/workspace/controllers/volume_controller.go:148 +0x2d8
github.com/onmetal/cephlet/controllers.(*VolumeReconciler).reconcileExists(0x0?, {0x1de04f8?, 0xc000c26570?}, {{0x1de2ff0?, 0xc000c265a0?}, 0xc000c26570?}, 0xc000979a80?)
	/workspace/controllers/volume_controller.go:121 +0x8a
github.com/onmetal/cephlet/controllers.(*VolumeReconciler).Reconcile(0xc00002a9c0, {0x1de04f8, 0xc000c26570}, {{{0xc000979a80?, 0x0?}, {0xc000979a96?, 0x40de47?}}})
	/workspace/controllers/volume_controller.go:114 +0x190
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1de04f8?, {0x1de04f8?, 0xc000c26570?}, {{{0xc000979a80?, 0x186dc40?}, {0xc000979a96?, 0x0?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:122 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000599a40, {0x1de0450, 0xc000339cc0}, {0x196a440?, 0xc000a2e300?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:323 +0x35f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000599a40, {0x1de0450, 0xc000339cc0})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:274 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:235 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:231 +0x587
```



**Additional context**
Add any other context about the problem here.

--- a/controllers/volume_controller.go
+++ b/controllers/volume_controller.go
@@ -145,8 +145,9 @@ func (r *VolumeReconciler) reconcile(ctx context.Context, log logr.Logger, volum
 		return ctrl.Result{}, err
 	}
 
-	if err := r.handleImagePopulation(ctx, log, volume); err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to initiate image population: %w", err)
+	waitUntilSnapshotReady, err := r.handleImagePopulation(ctx, log, volume)
+	if err != nil || waitUntilSnapshotReady {
+		return ctrl.Result{}, err
 	}
 
 	pvc, waitUntilClaimBound, err := r.applyPVC(ctx, log, volume)
@@ -342,24 +343,34 @@ func (r *VolumeReconciler) applyPVC(ctx context.Context, log logr.Logger, volume
 	log.V(3).Info("Volume provided.")
 	return pvc, false, nil
 }
-func (r *VolumeReconciler) handleImagePopulation(ctx context.Context, log logr.Logger, volume *storagev1alpha1.Volume) error {
+func (r *VolumeReconciler) handleImagePopulation(ctx context.Context, log logr.Logger, volume *storagev1alpha1.Volume) (bool, error) {
 	if volume.Spec.Image == "" {
-		return nil
+		return false, nil
 	}
 
 	snapshot := &snapshotv1.VolumeSnapshot{}
 	if err := r.Get(ctx, types.NamespacedName{Namespace: volume.Namespace, Name: GetSanitizedImageNameFromVolume(volume)}, snapshot); err != nil {
 		if !errors.IsNotFound(err) {
-			return fmt.Errorf("unable to get snapshot: %w", err)
+			return true, fmt.Errorf("unable to get snapshot: %w", err)
 		}
 		log.V(1).Info("Requested snapshot not found: create image pvc and snapshot it.")
-		return r.createSnapshot(ctx, log, volume)
+		return true, r.createSnapshot(ctx, log, volume)
 	}
 
-	volumeCapacity := volume.Spec.Resources[corev1alpha1.ResourceStorage]
-	if !pointer.BoolDeref(snapshot.Status.ReadyToUse, false) || snapshot.Status.RestoreSize == nil {
-		log.Info("Referenced snapshot not ready or RestoreSize not defined", "snapshotName", snapshot.Name)
-		return nil
+	if snapshot.Status == nil || !pointer.BoolDeref(snapshot.Status.ReadyToUse, false) {
+		log.Info("Referenced snapshot not ready", "snapshotName", snapshot.Name)
+		return true, nil
+	}
+
+	volumeCapacity, ok := volume.Spec.Resources[corev1alpha1.ResourceStorage]
+	if !ok {
+		log.Info("Volume resource storage not defined ", "snapshotName", snapshot.Name)
+		return true, fmt.Errorf("volume resource storage not defined")
+	}
+
+	if snapshot.Status.RestoreSize == nil {
+		log.Info("Snapshot RestoreSize not defined", "snapshotName", snapshot.Name)
+		return true, fmt.Errorf("snapshot restore size not defined")
 	}
 
 	volumeSizeBytes := volumeCapacity.Value()
@@ -368,7 +379,7 @@ func (r *VolumeReconciler) handleImagePopulation(ctx context.Context, log logr.L
 		r.Eventf(volume, corev1.EventTypeWarning, ReasonVolumeSizeToSmall, "Requested volume size %d is less than the size %d for the source snapshot %s", volumeSizeBytes, snapshot.Status.RestoreSize.Value(), snapshot.Name)
 	}
 
-	return nil
+	return false, nil
 }
 
 func getImageSize(ctx context.Context, volume *storagev1alpha1.Volume) (*resource.Quantity, error) {

